code/controllers/subsystem/timer.dm

```diff
@@ -1,90 +1,253 @@
+
 
 /datum/subsystem/timer/New()
 	processing = list()
 	hashes = list()
+
+	timer_id_dict = list()
+	timer_src_dict = list()
+
+	clienttime_timers = list()
+
 	NEW_SS_GLOBAL(SStimer)
 
 
 /datum/subsystem/timer/stat_entry(msg)
-	..("P:[processing.len]")
+	..("P:[length(processing)] H:[length(hashes)] C:[length(clienttime_timers)]")
 
 /datum/subsystem/timer/fire()
-	if(!processing.len)
-		can_fire = 0 //nothing to do, lets stop firing.
-		return
-	for(var/datum/timedevent/event in processing)
-		if(event.timeToRun <= world.time)
-			event.callback.InvokeAsync()
-			qdel(event)
-		if (MC_TICK_CHECK)
-			return
+	if (length(clienttime_timers))
+		for (var/thing in clienttime_timers)
+			var/datum/timedevent/event = thing
+			if (event.timeToRun <= REALTIMEOFDAY)
+				var/datum/callback/callback = event.callback
+				qdel(event) //this is right, we delete the event before running it so that stack overflows don't cause us to repeatively run the same event
+				callback.InvokeAsync()
+
+			if (MC_TICK_CHECK)
+				return
+
+	if (head_offset + (world.tick_lag * BUCKET_LEN) < world.time || length(src.buckets) != BUCKET_LEN || world.tick_lag != bucket_resolution)
+		refill_buckets()
+
+	var/list/buckets = src.buckets
+	while (practical_offset <= BUCKET_LEN && head_offset + (practical_offset*world.tick_lag) <= world.time && !MC_TICK_CHECK)
+		var/datum/timedevent/event = buckets[practical_offset]
+		while (event)
+			var/datum/callback/callback = event.callback
+			if (!callback)
+				CRASH("Invalid event: event.timeToRun=[event.timeToRun]||qdeleted(event)=[qdeleted(event)]||world.time=[world.time]||head_offset=[head_offset]||practical_offset=[practical_offset]")
+			qdel(event) //same as above
+			callback.InvokeAsync()
+			if (MC_TICK_CHECK)
+				return
+			event = buckets[practical_offset]
+
+		practical_offset++
+
+/datum/subsystem/timer/proc/refill_buckets()
+	sortTim(processing, /proc/cmp_timer)
+	src.buckets = new(BUCKET_LEN)
+	practical_offset = 1
+	bucket_resolution = world.tick_lag
+	var/list/buckets = src.buckets
+	var/new_offset
+	for (var/thing in processing)
+		var/datum/timedevent/event = thing
+		if (!event)
+			processing -= event
+			continue
+
+		if (isnull(new_offset))
+			new_offset = event.timeToRun
+
+		var/bucket_pos = round((event.timeToRun - new_offset) / world.tick_lag) + 1
+		if (bucket_pos > BUCKET_LEN)
+			break
+
+		var/datum/timedevent/bucket_head = buckets[bucket_pos]
+		if (!bucket_head)
+			buckets[bucket_pos] = event
+			continue
+
+		if (!bucket_head.prev)
+			bucket_head.prev = bucket_head
+		event.next = bucket_head
+		event.prev = bucket_head.prev
+		event.next.prev = event
+		event.prev.next = event
+	head_offset = new_offset
+
 
 /datum/subsystem/timer/Recover()
 	processing |= SStimer.processing
 	hashes |= SStimer.hashes
 
 /datum/timedevent
+	var/id
 	var/datum/callback/callback
 	var/timeToRun
-	var/id
+	var/datum/timerid
```

Comment:
oh ya, i forgot i scratched this idea.

i was gonna split this up between a internal datum, and a datum that left the subsystem for oop timers, but i realized that nobody is gonna use it, and worst case we just add more timer interfaces by id or implement it when it becomes useful.